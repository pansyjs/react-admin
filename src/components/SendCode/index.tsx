import React from 'react';
import { Button } from 'antd';
import { BaseButtonProps } from 'antd/es/button/button';
import { isPromise } from '@/utils/base';

export interface SendCodeProps extends BaseButtonProps {
  second?: number;
  initText?: string;
  runText?: string;
  resetText?: string;
  visible?: boolean;
  onEnd?: () => void;
  onGetCaptcha?: () => boolean | Promise<any>;
}

interface DefaultProps {
  readonly second: number;
  readonly initText: string;
  readonly runText: string;
  readonly resetText: string;
}

interface State {
  buttonText: string;
  runSecond: number;
  lastSecond: number;
  start: boolean;
}

// 发送验证码倒计时组件
class SendCode extends React.Component<SendCodeProps, State> {
  private timer: NodeJS.Timer;
  static defaultProps: DefaultProps = {
    second: 60,
    initText: '获取短信验证码',
    runText: '{%s}秒后重新获取',
    resetText: '重新获取验证码'
  };

  readonly state: State = {
    buttonText: this.props.initText,
    runSecond: this.props.second,
    lastSecond: 0,
    start: false
  };

  componentWillUnmount() {
    this.timeout();
  }

  // 按钮点击回调
  handleClick = (event) => {
    event.preventDefault();
    const { onGetCaptcha } = this.props;
    const result = onGetCaptcha ? onGetCaptcha() : null;

    if (!result === false) return;

    if (isPromise(result)) {
      result.then(this.start);
    } else {
      this.start();
    }
  };

  // 开始倒计时函数
  start() {
    this.setState({
      start: true
    });
    this.timer = setInterval(() => {
      const { lastSecond, runSecond } = this.state;
      let second = lastSecond ? lastSecond : runSecond;

      this.setState({
        buttonText: this.getButtonText(second),
        runSecond: runSecond - 1
      });
      second <= 0 && this.timeout();
    }, 1000);
  }

  // 倒计时结束处理函数
  timeout() {
    const { resetText, onEnd } = this.props;
    this.setState({
      buttonText: resetText,
      start: false
    });
    onEnd && onEnd();
    // 清除定时器
    clearInterval(this.timer);
  }

  /**
   * 获取按钮文本
   * @param second 当前的倒计时秒数
   */
  getButtonText = (second: number): string => {
    const { runText } = this.props;
    return runText.replace(/\{([^{]*?)%s(.*?)\}/g, second.toString());
  };

  render() {
    const {
      second,
      initText,
      resetText,
      runText,
      onGetCaptcha,
      onEnd,
      ...rest
    } = this.props;
    const { buttonText, start } = this.state;

    return (
      <Button onClick={this.handleClick} {...rest} disabled={start}>
        {buttonText}
      </Button>
    );
  }
}

export default SendCode;
