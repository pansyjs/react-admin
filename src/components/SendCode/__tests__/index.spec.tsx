import React from 'react';
import { shallow } from 'enzyme';
import SendCode from '../index';

describe('Component SendCode', () => {
  describe('without props', () => {
    const component = shallow(<SendCode />);

    it('should match a snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
