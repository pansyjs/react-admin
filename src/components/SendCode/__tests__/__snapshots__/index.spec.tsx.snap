// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component SendCode without props should match a snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SendCode
    initText="获取短信验证码"
    resetText="重新获取验证码"
    runText="{%s}秒后重新获取"
    second={60}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "block": false,
      "children": "获取短信验证码",
      "disabled": false,
      "ghost": false,
      "loading": false,
      "onClick": [Function],
      "prefixCls": "ant-btn",
    },
    "ref": null,
    "rendered": "获取短信验证码",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "block": false,
        "children": "获取短信验证码",
        "disabled": false,
        "ghost": false,
        "loading": false,
        "onClick": [Function],
        "prefixCls": "ant-btn",
      },
      "ref": null,
      "rendered": "获取短信验证码",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
